Console over websocket

jukkar opened the pull request:
This is implementation of zephyr console over a websocket. Tested with Firefox and wsdump.py command line tool which are able to send and receive data to/from the console.
There are generic websocket support patches in this PR so that compilation will work, only last two patches are related to ws-console implementation.

jukkar committed:
net: websocket: Initial support for server websocket

This commit creates a websocket library that can be used by
applications. The websocket library implements currently only
server role and it uses services provided by net-app API.
The library supports TLS if enabled in configuration file.

This also adds websocket calls to HTTP app server if websocket
connection is established.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: websocket echo-server application

This is a http(s) server that supports also websocket.
It sends back any data sent to it over a websocket.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: websocket: Add console support

Add console driver that allows console session to be transferred
over a websocket connection.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: websocket: Add console sample application

This sample application implements a web service that provides
zephyr console over websocket.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

dbkinder commented on samples/net/ws_console/README.rst:
```diff
@@ -0,0 +1,35 @@
+.. _websocket-server-sample:
```
whoops, just notice in https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/3894/3/tests that this document label is already in use (must be unique across all the documents)

dbkinder reviewed (changes requested):
doc label conflict

pfalcon commented:
@andrewboie : Adding you here. You asked about integrating printk into console subsystem lately. But there're bigger ("better") integration issues. So, we have subsys/console which supports only UART, but uses interrupt mode, and thus known to work. It also has POSIXy/ANSICy design. And we have drivers/console, which has adhoc design, known to be lossy (I recently tried to use bluetooth shell, which uses drivers/console/uart, and I couldn't use it, because it requires typing in long commands, and loses characters on clipboard paste from the host). But there now will be whole 3 of those! But they're really drivers - an app needs to configure (call) explicitly console drivers which it wants to work with.

So, a good task would be integrating those "console drivers" into "console subsystem". That would require substantial rewrite of those drivers (telnet/websocket), to work with generic POSIX/ANSIC model instead of Zephyr's adhoc one.

Hope this comment is useful.

pfalcon commented on subsys/shell/shell.c:
```diff
@@ -570,6 +573,9 @@ void shell_init(const char *str)
 	telnet_register_input(&avail_queue, &cmds_queue, completion);
 #endif
+#ifdef CONFIG_WS_CONSOLE
+	ws_register_input(&avail_queue, &cmds_queue, completion);
```
@andrewboie : Here's what I mean by "an app needs to call each console driver explicitly".

tbursztyka reviewed (approved):
seems fine to me

jukkar commented:
Rebased over #1271 and made sure this works ok.

jukkar committed:
net: app: Do not allow local unspecified address in client

Do not allow :: or ANY address in client when sending data.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
tests: net: websocket: Add tests for websocket

This tests websocket by creating a websocket support http server
and sending data to it and verifying the returned data is the same.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Updated to latest websocket code + Cmake changes.

Note that this PR contains some patches that are also found in #1271 
This is not to be merged to v1.10

jukkar commented:
This is already merged as part of other PR.
